### PR TITLE
expand docs

### PR DESCRIPTION
**Pull request by conradoplg**
WIP to show the overall structure and contents

Some highlights:

- I'm using "ZF FROST" to differentiate between the library and the protocol
- I stopped including the README.md as the book index because it's awkward to keep links working inside it if it's simultaneously a GitHub README, a Rust Doc home, and a mdbook home
- I want to revamp the "User Documentation" to maybe not include each README/DKG and just link to `docs.rs` for each crate

### CODE REVIEW

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -62,14 +62,15 @@ jobs:
           mdbook-version: '0.4.18'
 
       # TODO: actions-mdbook does not yet have an option to install mdbook-mermaid https://github.com/peaceiris/actions-mdbook/issues/426
-      - name: Install mdbook
+      - name: Install plugins
         run: |
           cargo install mdbook-mermaid
-          
+          cargo install mdbook-admonish
```
Comment (dconnolly) [inline .github/workflows/docs.yml:68]: 👍

File: book/book.toml [hunk 2]
```diff
@@ -3,4 +3,15 @@ authors = ["Zcash Foundation <frost@zfnd.org>"]
...
 multilingual = false
 src = "src"
-title = "The FROST Book"
+title = "The ZF FROST Book"
+
+[preprocessor]
+
+[preprocessor.admonish]
+command = "mdbook-admonish"
+assets_version = "2.0.1" # do not edit: managed by `mdbook-admonish install`
```
Comment (dconnolly) [inline book/book.toml:12]: 👍

File: book/src/frost.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+This explain the main concepts and flows of FROST in a generic manner. These
+are important to understand how to use the library, but rest assured that the
+[Tutorial](tutorial.md) will have more concrete information.
+
+FROST is a threshold signature scheme. It allows splitting a Schnorr private key
+into `n` shares for a threshold `t`, such that `t` (or more) participants can
+together generate a signature that can be validated by the matching public key.
+One important aspect is that the resulting signature is indistinguishable from a
+non-threshold signature from the point of view of signature verifiers.
```
Comment (dconnolly) [inline book/src/frost.md:11]:
```suggestion
FROST is a threshold signature scheme. It allows splitting a Schnorr signing key
into `n` shares for a threshold `t`, such that `t` (or more) participants can
together generate a signature that can be validated by the corresponding verifying
key. One important aspect is that the resulting signature is indistinguishable from a
non-threshold signature from the point of view of signature verifiers.
```

File: book/src/frost.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+## Key Generation
+
+There are two options for generating FROST key shares. In both cases, after the
+key generation procedure, each participant will get:
+
+- a **secret share**;
+- a **verifying share** (which can be used by other participants to verify the
+  signature shares they produce);
+- a **group public key**, which is the public key matching the private key that was
+  split into shares.
```
Comment (dconnolly) [inline book/src/frost.md:27]: We might want to use more words (vs different words) to clarify that verifying signature shares is different than verifying a final signature, or something

File: book/src/frost.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+### Trusted Dealer Generation
+
+An existing key (which can be freshly generated) is split into shares. It's the
+simplest approach, but it has the downside of requiring the entire key to exist
+in memory at some point in time, which may not be desired in high security
+applications. However, it is much simpler to set up. It requires an
+authenticated and confidential communication channel to distribute each share to
+their respective participants.
```
Comment (dconnolly) [inline book/src/frost.md:36]: We may link to our terminology page for these channel property definitions

Comment (dconnolly) [inline book/src/frost.md:36]:
```suggestion
applications. However, it is much simpler to set up. It requires an
[authenticated and confidential communication channel](https://frost.zfnd.org/terminology.html#peer-to-peer-channel) to distribute each share to
their respective participants.
```

File: book/src/frost.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+To learn how to do Trusted Dealer Generation with the ZF FROST library, see
+TODO.
+
+### Distribtuted Key Generation
+
+A two-round protocol after which each participant will have their share of
+the secret, without the secret being ever present in its entirety in any of the
+participant's memory. Its downside is that it require a [broadcast
```
Comment (dconnolly) [inline book/src/frost.md:45]:
```suggestion
the secret, without the secret being ever present in its entirety in any
participant's memory. Its downside is that it require a [broadcast
```

File: book/src/frost.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+channel](https://frost.zfnd.org/terminology.html#broadcast-channel) on top of
+authenticated and confidential communication channel between each pair of
+participants, which may be difficult to deploy in practice. See guidelines in
```
Comment (dconnolly) [inline book/src/frost.md:48]:
```suggestion
channel](https://frost.zfnd.org/terminology.html#broadcast-channel) as well as 
an authenticated and confidential communication channel between each pair of
participants, which may be difficult to deploy in practice. See guidelines in
```

File: book/src/frost.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+To learn how to do Distributed Key Generation with the ZF FROST
+library, see TODO.
+
+
+
+## Signing
+
+Signing with FROST starts with a Coordinator (which can be one of the
+share holders, or not) which selects the message to be signed and
+the participants that will generated the signature.
```
Comment (dconnolly) [inline book/src/frost.md:60]:
```suggestion
the participants that will generate the signature.
```

File: book/src/frost.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+For example, in 2-of-3 signing, if 3 participants are selected,
+them all 3 must produce signature shares in order for the Coordinator
+be able to produce the final signature. Of course, the Coordinator
+is still free to start the process with only 2 participants if they wish.
+```
+
+## Verifying
+
+Signature verification is carried out as normal, along with the signed message
+and the group public key as inputs.
```
Comment (dconnolly) [inline book/src/frost.md:84]:
```suggestion
Signature verification is carried out as normal with single-party signatures, 
along with the signed message and the group verifying key as inputs.
```

File: book/src/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
+# Tutorial
+
+The ZF FROST library consists of multiple crates. `frost-core` contains
```
Comment (dconnolly) [inline book/src/tutorial.md:3]:
```suggestion
The ZF FROST suite consists of multiple crates. `frost-core` contains
```

File: book/src/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+a generic implementation of the protocol, which can't be used directly
+without a concrete suite.
```
Comment (dconnolly) [inline book/src/tutorial.md:5]:
```suggestion
without a concrete instantiation.
```

File: book/src/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+To generate the key shares, the dealer calls
+[`generate_with_dealer()`](https://docs.rs/frost-ristretto255/latest/frost_ristretto255/keys/fn.generate_with_dealer.html).
+It returns a `HashMap` mapping the (automatically generated) `Identifier`s to
+their respective `SecretShare`s, and a `PublicKeyPackage` which contains the
+`VerifyingShare` for each participant and the group public key (`VerifyingKey`).
+
+TODO: insert code snippet
+
+Each `SecretShare` must then be sent via a **authenticated** and
+**confidential** channel for each participant, who verify the
```
Comment (dconnolly) [inline book/src/tutorial.md:40]:
```suggestion
Each `SecretShare` must then be sent via an [**authenticated** and
**confidential** channel ](https://frost.zfnd.org/terminology.html#peer-to-peer-channel)for each participant, who verify the
```

File: book/src/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+package to obtain a `KeyPackage` which contains their secret share,
```
Comment (dconnolly) [inline book/src/tutorial.md:41]:
```suggestion
package to obtain a `KeyPackage` which contains their signing share,
```

File: book/src/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+```admonish danger
+Which **authenticated** and **confidential** channel to use is up to the
+application. Some examples:
+
+- Manually require the dealer to sent the `SecretShare`s to the
+  partipants using some secure messenger such as Signal;
+- Use a TLS connection, authenticating the server with a certificate
+  and the client with some user/password or another suitable authentication
+  mechanism;
+
```
Comment (dconnolly) [inline book/src/tutorial.md:69]: This duplicates a little bit of the peer-to-peer section in the terminology page, which I'm actually ok with as it's not easy to just inline that section fully in here, but we can link too

File: book/src/tutorial.md [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+Which encoding to use to transmit the `SecretShare` is up for the developer.
+Note that specific types (Scalars, Elements) do have a standardized encoding.
+Also note that `serde` support will be added in the future to offer a default
+method for encoding.
+
+The same applies to `KeyPackage`.
+```
+
+```admonish danger
+Which **authenticated** and **confidential** channel to use is up to the
```
Comment (dconnolly) [inline book/src/tutorial.md:61]:
```suggestion
Which [**authenticated** and **confidential** channel](https://frost.zfnd.org/terminology.html#peer-to-peer-channel) to use is up to the
```

File: book/src/tutorial.md [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+- Manually require the dealer to sent the `SecretShare`s to the
+  partipants using some secure messenger such as Signal;
+- Use a TLS connection, authenticating the server with a certificate
+  and the client with some user/password or another suitable authentication
+  mechanism;
+
+Failure of using a **confidential** channel may lead to the shares being
+stolen and possibly allowing signature forgeries if a threshold of them
+are stolen.
```
Comment (dconnolly) [inline book/src/tutorial.md:72]:
```suggestion
Failure of using a **confidential** channel may lead to the shares being
stolen and possibly allowing signature forgeries if a threshold number of 
them are stolen.
```

File: book/src/tutorial.md [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+```admonish danger
+The `SecretPackage` contents must be stored securely. For example:
+
+- Make sure other users in the system can't read it;
+- If possible, use the OS secure storage such that the package
+  contents can only be opened with the user's password or biometrics.
+```
+
+```admonish warning
+The participants may wish to not fully trust the dealer. While **the dealer
```
Comment (dconnolly) [inline book/src/tutorial.md:88]:
```suggestion
The participants may wish to not fully trust the dealer. **The dealer
```

**Comment by dconnolly** [comment]
Looking good!

File: book/src/tutorial.md (hunk 14 above)
Comment (conradoplg) [reply book/src/tutorial.md]: Good point, you added a link but I also added a more explicit reference in 5d55082

File: book/src/frost.md (hunk 4 above)
Comment (conradoplg) [reply book/src/frost.md]: Done in 5d55082

**Comment by dconnolly** [comment]
gencode code needs an update? 
image

**Comment by conradoplg** [comment]
> gencode code needs an update? image

Oops. Fixed

**Review by mpguerra — COMMENTED** [review]
Looks great! I specially love the warnings explaining how things might be done in a certain way and the consequences of not sharing various different artifacts in the correct way.

I did an editorial review for language and content without looking into the correctness of the text. There are some minor readability suggestions.

**Review by mpguerra — APPROVED** [review]

**Pull request merged**